docs: Add a chapter on start states to the book

ratmice opened the pull request:
Despite the fact that I am the terriblest writer, here is a rough draft for a chapter on *start states*.

I still need to test the code examples and stuff so marking as draft, but would appreciate any feedback.

ltratt commented on doc/src/start_states.md:
```diff
@@ -0,0 +1,118 @@
+The following explains the syntax and semantics of Start States in lrlex.<br>
+A working example can be found in the repository at [lrpar/examples/start_states](https://github.com/softdevteam/grmtools/tree/master/lrpar/examples/start_states)
+
+# Motivation
```
Minor comment: `#` is really only for the top-level comment in an MD file (so the file should probably start with the line `# Start states`) and then these comments should be `##`.

ratmice replied:
Should be fixed in [3fad471](https://github.com/softdevteam/grmtools/pull/468/commits/3fad4717f20cdb9722ca715e2cb6fb4cec665894)

ltratt commented on doc/src/start_states.md:
```diff
@@ -0,0 +1,118 @@
+
+# Motivation
+
+Start states are a feature from lex which can be used to add a bit of context sensitivity to lexical analysis.
```
"which can be used for context sensitive lexing" maybe?

ratmice replied:
Should be fixed in [d182e4a](https://github.com/softdevteam/grmtools/pull/468/commits/d182e4a8897848d1d6935e26471a13635665c3e8)

ltratt commented on doc/src/start_states.md:
```diff
@@ -0,0 +1,118 @@
+Additionally the lexer has a stack of states, and matching rules perform actions which modify the stack.
+
+# The INITIAL start state 
+By default unless specified otherwise all lex rules are members of the *INITIAL* start state.
```
Can drop "By default"?

ratmice replied:
Should be fixed in [d182e4a](https://github.com/softdevteam/grmtools/pull/468/commits/d182e4a8897848d1d6935e26471a13635665c3e8)

ltratt commented on doc/src/start_states.md:
````diff
@@ -0,0 +1,118 @@
+<INITIAL>[\t \n]+ ;
+```
+
+Being equivalent to the lex file below with no start states specified.
````
s/equivalent to/This is/

ratmice replied:
Should be fixed in [d182e4a](https://github.com/softdevteam/grmtools/pull/468/commits/d182e4a8897848d1d6935e26471a13635665c3e8)

ltratt commented on doc/src/start_states.md:
````diff
@@ -0,0 +1,118 @@
+<Excl>=# <-Excl>;
+```
+
+# Start state operators
````
I wonder if this section should come earlier and be called something like "Differences from POSiX lex"?

ratmice replied:
I moved this one section higher, Which I believe does make sense because the section higher had start state operators in it, and the section renaming did seem appropriate too.

Should be fixed in [aaaa782](https://github.com/softdevteam/grmtools/pull/468/commits/aaaa78201d0a7158a8b5d76fb2e18f2c0d45e901)

ltratt commented:
This all looks good, and worth merging now. Can you line wrap to the same length as the other md files (I think 100 chars) and force push?

ratmice commented:
Will try and do so later today, as well as going through and running all the code snippets.

ratmice commented:
Because I made some additional changes (changed the link format so it would fit on 100 char lines), I was hesitant to squash/force push.

I tested all the *complete* examples containing e.g. those containing `%%` and fixed the one error I found.
Also removed a line that was no longer relevant.

I am happy with this now, and feel like it's ready for a squash.

ltratt commented:
LGTM! Please squash.

ratmice commented:
Squashed.
